<a href="https://colab.research.google.com/github/Lexi-Zhou/stats201-project-zzz/blob/main/Code/W2_1_Data_Preparation.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# STATS 201 Final Project
## Week 2 Dataset initional exploratory analysis

- Datasource: https://data.mendeley.com/datasets/fvtfjyvw7d/1


## 1. Data preparation:


### 1.1 Combine datasets
Skip after having merged the datasets into one.

In [ ]:
# Load dataset
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
# folder_path = "/content/drive/MyDrive/Colab Notebooks/stats201_final_project/Week_2/RMP_Dataset"


In [ ]:
import pandas as pd
import glob
import os

In [ ]:
# csv_files = glob.glob(os.path.join(folder_path, "*.csv"))

# len(csv_files)

2770

In [ ]:
# Merge datasets
output_path = "/content/drive/MyDrive/Colab Notebooks/stats201_final_project/Week_2/RMP_merged.csv"

# first = True

# for file in csv_files:
#    df = pd.read_csv(file)
#    df.to_csv(
#        output_path,
#        mode="w" if first else "a",
#        header=first,
#        index=False
#   )
#    first = False


In [ ]:
# sanity check

# merged_df = pd.read_csv(output_path)
# merged_df.shape
# merged_df.head()
# merged_df.info()


In [ ]:
# first = True

# for file in csv_files:
#    df = pd.read_csv(file)
#    df["source_file"] = os.path.basename(file)

#    df.to_csv(
#        output_path,
#        mode="w" if first else "a",
#        header=first,
#        index=False
#    )
#    first = False


#### 1.1.1 Dataset Variable Description

| Column Name | Data Type | Description |
|------------|----------|-------------|
| professor_name | Text | Full name of the professor being reviewed. |
| school_name | Text | Name of the college or university where the professor teaches. |
| department_name | Text | Academic department to which the professor belongs. |
| local_name | Text | City or region of the institution. |
| state_name | Text | State abbreviation of the institution location. |
| year_since_first_review | Numeric (Integer) | Number of years since the professor received their first review on the platform. |
| star_rating | Numeric (Float) | Overall average rating of the professor. |
| take_again | Numeric / Missing | Percentage or score indicating how likely students would take the professor again. |
| diff_index | Numeric (Float) | Professor-level average of student-reported course difficulty ratings. |
| tag_professor | Text | Aggregated tags associated with the professor, including tag frequency counts. |
| num_student | Numeric (Integer) | Total number of students who have submitted reviews for the professor. |
| post_date | Date | Date when the individual student review was posted. |
| name_onlines | Text | Course code or course identifier associated with the review. |
| student_star | Numeric (Float) | Star rating given by an individual student for a specific course (1.0 - 5.0 in 0.5 increments). |
| student_difficult | Numeric (Float) | Student-reported difficulty level of the course (1.0 - 5.0 in 0.5 increments). |
| attence | Text | Indicates whether attendance was mandatory for the course. |
| for_credits | Text | Indicates whether the course was taken for credit. |
| would_take_agains | Text | Student's self-reported willingness to take the professor again. |
| grades | Text | Grade reported by the student for the course. |
| stu_tags | Text | Tags selected by the student to describe the professor or course. |
| help_useful | Numeric (Integer) | Number of users who marked the review as helpful. |
| help_not_useful | Numeric (Integer) | Number of users who marked the review as not helpful. |
| comments | Text | Free-text student comment describing their experience with the professor and course. |


### 1.2 Select and standardize columns

In [ ]:
## read dataset

import pandas as pd

RMP_merged = pd.read_csv(
    "/content/drive/MyDrive/Colab Notebooks/stats201_final_project/Week_2/RMP_merged.csv"
)



In [ ]:
RMP_merged.head()

professor_name                      school_name       department_name  \
0   Lynn  Ketter  Northwest Florida State College  Education department   
1   Lynn  Ketter  Northwest Florida State College  Education department   
2   Lynn  Ketter  Northwest Florida State College  Education department   
3   Lynn  Ketter  Northwest Florida State College  Education department   
4   Lynn  Ketter  Northwest Florida State College  Education department   

   local_name state_name  year_since_first_review  star_rating take_again  \
0   Niceville         FL                      9.0          3.7        NaN   
1   Niceville         FL                      9.0          3.7        NaN   
2   Niceville         FL                      9.0          3.7        NaN   
3   Niceville         FL                      9.0          3.7        NaN   
4   Niceville         FL                      9.0          3.7        NaN   

   diff_index                                      tag_professor  ...  \
0         2.5  Gives good feedback (3)  Participation matters...  ...   
1         2.5  Gives good feedback (3)  Participation matters...  ...   
2         2.5  Gives good feedback (3)  Participation matters...  ...   
3         2.5  Gives good feedback (3)  Participation matters...  ...   
4         2.5  Gives good feedback (3)  Participation matters...  ...   

   student_star student_difficult    attence  for_credits  would_take_agains  \
0           5.0               2.0  Mandatory          Yes                Yes   
1           2.0               2.0        NaN          Yes                 No   
2           3.5               3.0  Mandatory          Yes                NaN   
3           4.0               3.0        NaN          Yes                NaN   
4           5.0               2.0        NaN          NaN                NaN   

  grades                                           stu_tags help_useful  \
0     A+             RESPECTED  CARING  GIVES GOOD FEEDBACK           0   
1      A                              PARTICIPATION MATTERS           0   
2      B  GIVES GOOD FEEDBACK  LOTS OF HOMEWORK  PARTICI...           0   
3    NaN  GIVES GOOD FEEDBACK  LOTS OF HOMEWORK  SKIP CL...           0   
4    NaN                                                NaN           0   

  help_not_useful                                           comments  
0               0  Wonderful teacher. Basically all work is done ...  
1               0  Discussion boards are mandatory. Use book for ...  
2               0  Honestly, I didn\'t learn anything in this cla...  
3               0  She comes off a little cold at first but she\'...  
4               0  I thought she was a great teacher and always g...  

[5 rows x 23 columns]

In [ ]:
# Standardize department names (delete " department")

RMP_merged["department_name_clean"] = (
    RMP_merged["department_name"]
    .str.replace(r"\s*department$", "", regex=True)
)



## department distribution
dept_counts = (
    RMP_merged["department_name_clean"]
    .value_counts()
    .head(20)
)

dept_counts


department_name_clean
English              2238
Mathematics          1687
Psychology           1052
Biology               982
Chemistry             884
History               678
Business              645
Sociology             457
Education             403
Accounting            401
Computer Science      397
Communication         396
Languages             382
Nursing               364
Economics             361
Philosophy            352
Music                 339
Political Science     334
Engineering           321
Law                   275
Name: count, dtype: int64

In [ ]:
# Keep only the columns needed for analysis

needed_cols = ["professor_name", "school_name", "department_name_clean", "stu_tags", "student_star", "comments"]


In [ ]:
df = RMP_merged[needed_cols]

professor_name                      school_name  \
0          Lynn  Ketter  Northwest Florida State College   
1          Lynn  Ketter  Northwest Florida State College   
2          Lynn  Ketter  Northwest Florida State College   
3          Lynn  Ketter  Northwest Florida State College   
4          Lynn  Ketter  Northwest Florida State College   
...                 ...                              ...   
19680        Jon  Sands         Arizona State University   
19681  Randolph  Haines         Arizona State University   
19682  Randolph  Haines         Arizona State University   
19683  Randolph  Haines         Arizona State University   
19684     Howard  Cabot         Arizona State University   

      department_name_clean  \
0                 Education   
1                 Education   
2                 Education   
3                 Education   
4                 Education   
...                     ...   
19680                   Law   
19681                   Law   
19682                   Law   
19683                   Law   
19684                   Law   

                                                stu_tags  student_star  \
0                 RESPECTED  CARING  GIVES GOOD FEEDBACK           5.0   
1                                  PARTICIPATION MATTERS           2.0   
2      GIVES GOOD FEEDBACK  LOTS OF HOMEWORK  PARTICI...           3.5   
3      GIVES GOOD FEEDBACK  LOTS OF HOMEWORK  SKIP CL...           4.0   
4                                                    NaN           5.0   
...                                                  ...           ...   
19680                                                NaN           5.0   
19681                                                NaN           3.5   
19682                                                NaN           4.0   
19683                                                NaN           2.0   
19684                                                NaN           4.0   

                                                comments  
0      Wonderful teacher. Basically all work is done ...  
1      Discussion boards are mandatory. Use book for ...  
2      Honestly, I didn\'t learn anything in this cla...  
3      She comes off a little cold at first but she\'...  
4      I thought she was a great teacher and always g...  
...                                                  ...  
19680                           Great at oral advocacy!!  
19681  Prof (Judge) Haines is a great Prof. He is cha...  
19682  This is a very highly intelligent man who expe...  
19683  Enrolled in Ch. 11 without being told their wa...  
19684  Weird class. Lots of reading, but you probably...  

[19685 rows x 6 columns]

In [ ]:
# trim whitespace in key text columns to reduce accidental duplicates.
for c in ["professor_name", "school_name", "department_name_clean"]:
    df[c] = df[c].astype(str).str.strip()


/tmp/ipython-input-258290273.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[c] = df[c].astype(str).str.strip()
/tmp/ipython-input-258290273.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df[c] = df[c].astype(str).str.strip()
/tmp/ipython-input-258290273.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_gu

### 1.3 Categorize each entry by rating group

Rule:
- negative: 1.0 to 2.5 (inclusive)
- neutral: exactly 3.0
- positive: 3.5 to 5.0 (inclusive)

In [ ]:
def rating_group(x: float) -> str:
  if pd.isna(x):
    return "missing"
  if 1.0 <= x <= 2.5:
    return "negative"
  if x == 3.0:
    return "neutral"
  if 3.5 <= x <= 5.0:
    return "positive"
  return "other"


In [ ]:
df["rating_group"] = df["student_star"].apply(rating_group)

/tmp/ipython-input-2294167362.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["rating_group"] = df["student_star"].apply(rating_group)


In [ ]:
df

professor_name                      school_name  \
0              Lynn  Ketter  Northwest Florida State College   
1              Lynn  Ketter  Northwest Florida State College   
2              Lynn  Ketter  Northwest Florida State College   
3              Lynn  Ketter  Northwest Florida State College   
4              Lynn  Ketter  Northwest Florida State College   
..                      ...                              ...   
189  Gloria  Depew-Kendrick      Schoolcraft College - (ALL)   
190  Gloria  Depew-Kendrick      Schoolcraft College - (ALL)   
191  Gloria  Depew-Kendrick      Schoolcraft College - (ALL)   
192  Gloria  Depew-Kendrick      Schoolcraft College - (ALL)   
193  Gloria  Depew-Kendrick      Schoolcraft College - (ALL)   

    department_name_clean                                           stu_tags  \
0               Education             RESPECTED  CARING  GIVES GOOD FEEDBACK   
1               Education                              PARTICIPATION MATTERS   
2               Education  GIVES GOOD FEEDBACK  LOTS OF HOMEWORK  PARTICI...   
3               Education  GIVES GOOD FEEDBACK  LOTS OF HOMEWORK  SKIP CL...   
4               Education                                                NaN   
..                    ...                                                ...   
189        Health Science                                                NaN   
190        Health Science                                                NaN   
191        Health Science                                                NaN   
192        Health Science                                                NaN   
193        Health Science                                                NaN   

     student_star                                           comments  \
0             5.0  Wonderful teacher. Basically all work is done ...   
1             2.0  Discussion boards are mandatory. Use book for ...   
2             3.5  Honestly, I didn\'t learn anything in this cla...   
3             4.0  She comes off a little cold at first but she\'...   
4             5.0  I thought she was a great teacher and always g...   
..            ...                                                ...   
189           5.0  She is great at what she does, Teaching! And t...   
190           5.0  Gloria is awesome. I would highly recommend ta...   
191           5.0  Gloria is a terrific medical terminology instr...   
192           5.0  Gloria is terrific!! Highly recommend her. She...   
193           5.0  Absolutely Wonderful Teacher! I would reccomme...   

    rating_group  
0       positive  
1       negative  
2       positive  
3       positive  
4       positive  
..           ...  
189     positive  
190     positive  
191     positive  
192     positive  
193     positive  

[194 rows x 7 columns]

### 1.4 Parse student tags into unique tag labels

In RMP, stu_tags are often phrases separated by TWO OR MORE spaces, not single spaces.

Example: "TEST HEAVY  LECTURE HEAVY  SKIP CLASS? YOU WON'T PASS."

Here each tag is a multi-word phrase, and tags are separated by "  " (double spaces).

We split on 2+ whitespace characters to keep multi-word tags intact.


In [ ]:
import re

def split_tags(tag_str: str) -> list:
  if pd.isna(tag_str):
    return []
  s = str(tag_str).strip()
  if not s:
    return []
  # split tags by two or more whitespace characters (robust to variable spacing)
  parts = re.split(r"\s{2,}", s)
  # clean each tag phrase: strip and drop empties
  parts = [p.strip() for p in parts if p and p.strip()]
  return parts

df["tag_list"] = df["stu_tags"].apply(split_tags)

/tmp/ipython-input-4083366252.py:15: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df["tag_list"] = df["stu_tags"].apply(split_tags)


In [ ]:
# Explode tag_list to one tag per row (for counting and uniques).
tags_long = df[["rating_group", "tag_list"]].explode("tag_list").rename(columns={"tag_list": "tag_label"})
tags_long["tag_label"] = tags_long["tag_label"].fillna("").astype(str).str.strip()
tags_long = tags_long[tags_long["tag_label"] != ""]  # remove empty tags


In [ ]:
# sample

print(RMP_merged.shape)
print(RMP_merged["professor_name"].nunique())
print(RMP_merged["school_name"].nunique())
print(RMP_merged["department_name"].nunique())


(19685, 24)
2766
1127
270


### 1.5 Summary tables


#### 1.5.1 Overall uniques

In [ ]:
overall_uniques = pd.DataFrame([{
   "n_rows": len(df),
    "unique_professors": df["professor_name"].nunique(dropna=True),
    "unique_schools": df["school_name"].nunique(dropna=True),
    "unique_departments": df["department_name_clean"].nunique(dropna=True),
    "unique_tag_labels": tags_long["tag_label"].nunique(dropna=True),
    "stu_star_nonmissing": df["student_star"].notna().sum(),
    "stu_star_mean": df["student_star"].mean(),
    "stu_star_median": df["student_star"].median(),

}])

##### 1.5.2 Group-level summary by rating group

In [ ]:
group_uniques = (
    df.groupby("rating_group", dropna=False)
      .agg(
          n_rows=("student_star", "size"),
          unique_professors=("professor_name", "nunique"),
          unique_schools=("school_name", "nunique"),
          unique_departments=("department_name_clean", "nunique"),
          stu_star_nonmissing=("student_star", lambda x: x.notna().sum()),
          stu_star_mean=("student_star", "mean"),
          stu_star_median=("student_star", "median"),
          stu_star_min=("student_star", "min"),
          stu_star_max=("student_star", "max"),
      )
      .reset_index()
)

In [ ]:
# Add "unique_tag_labels" per group using the exploded tags table.
tag_uniques_by_group = (
    tags_long.groupby("rating_group")["tag_label"]
            .nunique()
            .reset_index()
            .rename(columns={"tag_label": "unique_tag_labels"})
)

group_uniques = group_uniques.merge(tag_uniques_by_group, on="rating_group", how="left")
group_uniques["unique_tag_labels"] = group_uniques["unique_tag_labels"].fillna(0).astype(int)


In [ ]:
display(overall_uniques)
display(group_uniques)

n_rows  unique_professors  unique_schools  unique_departments  \
0   19685               2766            1127                 270   

   unique_tag_labels  stu_star_nonmissing  stu_star_mean  stu_star_median  
0                 20                19672       3.766241              4.5

rating_group  n_rows  unique_professors  unique_schools  unique_departments  \
0      missing      13                  9               9                   7   
1     negative    4862               1544             807                 213   
2      neutral    1195                802             490                 152   
3     positive   13615               2481            1064                 251   

   stu_star_nonmissing  stu_star_mean  stu_star_median  stu_star_min  \
0                    0            NaN              NaN           NaN   
1                 4862       1.561394              1.5           1.0   
2                 1195       3.000000              3.0           3.0   
3                13615       4.620859              5.0           3.5   

   stu_star_max  unique_tag_labels  
0           NaN                  0  
1           2.5                 20  
2           3.0                 20  
3           5.0                 20

**Descriptive Analysis of Rating Groups**

We categorized each student comment into three rating groups based on the `student_star` score: **negative** (1.0-2.5), **neutral** (3.0), and **positive** (3.5-5.0). Out of 19,685 total comments, the vast majority (13,615, \~69\%) fall into the positive group, reflecting generally high student satisfaction. The negative group comprises 4,862 comments (\~25\%), while the neutral group is the smallest with 1,195 comments (\~6\%).

Comments in the positive group are associated with 2,481 unique professors across 1,064 schools and 251 unique tag labels, and exhibit a mean `student_star` of 4.62 and median of 5.0. The negative group includes 1,544 professors from 807 schools, with a mean rating of 1.56 and median of 1.5, indicating strong dissatisfaction. The neutral group, with mean and median exactly at 3.0, spans 802 professors and 490 schools.

This distribution indicates a pronounced skew toward positive evaluations, but substantial variability exists across professors and schools. Notably, the negative and neutral groups still cover a wide range of departments and tag labels, suggesting that even lower-rated comments are spread across diverse academic contexts rather than being isolated to a few instructors or courses. This preliminary categorization will serve as the basis for further analysis of text features and gendered perception in student evaluations.

In [ ]:
## save outputs

# Save the prepared analysis dataset (selected columns + rating_group + tag_list).
out_data_path = "/content/drive/MyDrive/Colab Notebooks/stats201_final_project/Week_2/RMP_analysis_ready.csv"
df.to_csv(out_data_path, index=False)

# Save summary tables.
out_overall_path = "/content/drive/MyDrive/Colab Notebooks/stats201_final_project/Week_2/RMP_summary_overall_uniques.csv"
out_group_path = "/content/drive/MyDrive/Colab Notebooks/stats201_final_project/Week_2/RMP_summary_by_rating_group.csv"
overall_uniques.to_csv(out_overall_path, index=False)
group_uniques.to_csv(out_group_path, index=False)